[CPU] [IE TESTS] Updated Pad

a-sidorova opened the pull request:
This is the continuation of this [PR](https://github.com/openvinotoolkit/openvino/pull/2586).

Issues: 27177, 38891, 42082, 30586, 38226

Changes:
- Added blocked layouts support (for Constant padsBegin[1] and padsEnd[1] / 8[16], for others modes channel paddings are 0)
- Added support `nhwc` and `ndhwc` formats
- Added support for different data types
- Added tests for blocked layouts and `n[d]hwc`
- Added more optimal implementations
- Disabled `ConvertPadToGroupConvolution` in CPU plugin

dmitry-gorokhov commented:
@maxnick  Please review the PR

maxnick commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -95,158 +203,327 @@ void MKLDNNPadNode::createPrimitive() {
+                    break;
+                case InferenceEngine::Precision::U8:
+                    padConstant<uint8_t>();
+                    break;
```
Please add BF16 type support.

maxnick commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -95,158 +203,327 @@ void MKLDNNPadNode::createPrimitive() {
+}
 
+void MKLDNNPadNode::execute(mkldnn::stream strm) {
     switch (padMode) {
```
I think it would be better to rewrite the switch statement using the Conditional Compilation library.

maxnick commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -95,158 +203,327 @@ void MKLDNNPadNode::createPrimitive() {
+
+            InferenceEngine::Precision precision = this->getSelectedPrimitiveDescriptor()->getConfig().inConfs[0].desc.getPrecision();
+            switch (precision) {
+                case InferenceEngine::Precision::FP32:
```
Please use the Conditional Compilation library.

maxnick commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.h:
```diff
@@ -29,26 +30,36 @@ class MKLDNNPadNode : public MKLDNNNode {
-    size_t getWorkAmountDst() const;
+    void prepareParams();
+
+    inline size_t parallel_init(size_t start, std::vector<size_t> &indexes);
```
It is better to rewrite all such auxiliary functions as `static inline` and remove from the header.

maxnick commented on inference-engine/tests/functional/plugin/cpu/single_layer_tests/pad.cpp:
```diff
@@ -0,0 +1,297 @@
+
+
+const std::vector<InferenceEngine::Precision> inputPrecisions = {
+        InferenceEngine::Precision::FP32,
```
FP16 seems to be useless as the MKDLDNN plugin converts it to FP32. But it would be better to add another precisions natively supported by the MKLDNN plugin.

a-sidorova commented:
@maxnick, thank you for review!

I fixed all, added special paddings for Const mode tests with blocked layouts, added BF16 support and used the Conditional Compilation library for padConstant.

maxnick commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -95,158 +203,312 @@ void MKLDNNPadNode::createPrimitive() {
-inline size_t parallel_init(size_t start, size_t size, std::vector<size_t> &counters, std::vector<size_t> &dims) {
-    for (int j = size - 1; j >= 0; j--) {
-        counters[j] = start % dims[j];
+inline size_t parallel_init(size_t start, size_t nDims, const SizeVector& dims, SizeVector& indexes) {
```
Please specify such functions as `static` to restrict their scope.

a-sidorova replied:
Thanks, done

maxnick commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.h:
```diff
@@ -29,27 +32,44 @@ class MKLDNNPadNode : public MKLDNNNode {
+    template<typename T> friend struct PadConstantEmitter;
+};
+
+template<typename T>
```
This struct is already forward declared, so it could be moved to the cpp.

a-sidorova replied:
Discussed offline

maxnick commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.h:
```diff
@@ -29,27 +32,44 @@ class MKLDNNPadNode : public MKLDNNNode {
+    void prepareParams();
 
-    size_t getWorkAmountDst() const;
+    inline void getDstIdx(const InferenceEngine::SizeVector& indexes, size_t& dstIdx) const;
```
This fuctions although could be moved to the cpp as `static inline`.

a-sidorova replied:
Discussed offline. This method accesses class fields

maxnick commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.h:
```diff
@@ -29,27 +32,44 @@ class MKLDNNPadNode : public MKLDNNNode {
-    void padReflectOrSymmetric(const float *srcData, float* dstData,
-                               InferenceEngine::SizeVector srcDims, InferenceEngine::SizeVector dstDims,
-                               const bool isSymmetric = false);
+    void padConstant();
```
It is better to place such auxiliary functions, which are not a part of the class interface, to the private section.

maxnick commented on inference-engine/tests/functional/plugin/cpu/single_layer_tests/pad.cpp:
```diff
@@ -0,0 +1,304 @@
+
+
+const auto cpuParams_nChw16c = CPUSpecificParams {{nChw16c}, {nChw16c}, {}, "ref_FP32"};
+const auto cpuParams_nCdhw16c = CPUSpecificParams {{nCdhw16c}, {nCdhw16c}, {}, "ref_FP32"};
```
`"ref_FP32"` won't work with BF16, please add `selectedType` precision adjustment to the SeyUp routine.

a-sidorova replied:
Thank you, done!

maxnick commented on inference-engine/tests/functional/plugin/cpu/single_layer_tests/pad.cpp:
```diff
@@ -0,0 +1,304 @@
+const std::vector<InferenceEngine::Precision> inputPrecisions = {
+        InferenceEngine::Precision::FP32,
+        InferenceEngine::Precision::FP16,
+        InferenceEngine::Precision::BF16
```
Please add I8 to cover all supported data sizes.

a-sidorova replied:
Thanks, I added I8

maxnick reviewed (approved):
In general LGTM.

a-sidorova commented:
@maxnick, thank you!

@dmitry-gorokhov, could you review please?

dmitry-gorokhov commented:
@a-sidorova Please resolve comments if they have been fixed.

dmitry-gorokhov commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -65,10 +67,11 @@ void MKLDNNPadNode::initSupportedPrimitiveDescriptors() {
         return;
 
-    InferenceEngine::Precision precision = InferenceEngine::Precision::FP32;
+    InferenceEngine::Precision precision = getCnnLayer()->insData[0].lock()->getPrecision();
```
This change is not always valid: for example your implementation doesn't support BIN precision. You need to create list of supported by your implementation precisions and check if getCnnLayer()->insData[0].lock()->getPrecision() is present in the list. Otherwise you need to create supportedPrimitiveDescriptors for FP32 precision in case getCnnLayer()->insData[0].lock()->getPrecision().is_float() == true and I32 if false. The plug-in will automatically do all necessary data type conversions.

a-sidorova replied:
Its my mistake. Thank you! Fixed

dmitry-gorokhov replied:
BTW, implementation limitation on precisions is true only for Constant mode with non zero const value.  It seems worthwhile to check precisions only for this case.

dmitry-gorokhov commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -95,158 +202,312 @@ void MKLDNNPadNode::createPrimitive() {
+
+template<typename T>
+void MKLDNNPadNode::padConstantCommon() {
+    const T *srcData = getParentEdgeAt(0)->getBlob()->cbuffer().as<const T*>() +
```
getDataPtr?

a-sidorova replied:
We can't use this method, because for better perfomance when padValue != 0 we should use `std::fill_n` which doesn't support `uint8_t*` in `getDataPtr`. When you want to use this method, you must use `cpu_memcpy`, not `std::fill_n`. But its slower. So I preferred `std::fill_n` and `getBlob()->cbuffer()`

dmitry-gorokhov replied:
reinterpret_cast<T*>(getDataPtr(...))?

dmitry-gorokhov commented on inference-engine/tests/functional/plugin/cpu/single_layer_tests/pad.cpp:
```diff
@@ -0,0 +1,305 @@
+        testing::Values(InferenceEngine::Precision::UNSPECIFIED),
+        testing::Values(InferenceEngine::Precision::UNSPECIFIED),
+        testing::Values(InferenceEngine::Layout::ANY),
+        testing::Values(std::vector<size_t>{3, 16, 5, 5}),
```
It would be good to check cases when number of channels % block_size != 0

a-sidorova replied:
I didn't add paddings for channels % block_size != 0 case for Constant mode because I must use useless `if else` for planar layouts. So it would be slower. And I didn't see cases when we must use blocked layouts with paddings for channel % block_size != 0.

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -82,7 +85,112 @@ void MKLDNNPadNode::initSupportedPrimitiveDescriptors() {
+                    selectPrimitiveDescriptorByIndex(pd);
+                    return;
+                }
+            } else if (desc.getBlockingDesc().getBlockDims()[nLastDim] == 16) {
```
Thanks, done

a-sidorova commented on inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:
```diff
@@ -203,8 +203,8 @@ static void Transformation(ICNNNetwork::Ptr& clonedNetwork, const Config& conf)
     pass_config->disable<ngraph::pass::ConvertMod>();
     pass_config->disable<ngraph::pass::LogSoftmaxDecomposition>();
     pass_config->disable<ngraph::pass::ConvertInterpolateToInterpOrResampleMatcher>();
+    pass_config->disable<ngraph::pass::ConvertPadToGroupConvolution>();
```
Ok, done

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -82,7 +85,111 @@ void MKLDNNPadNode::initSupportedPrimitiveDescriptors() {
+        memoryFormat = mkldnn::memory::nhwc;
+        config.inConfs[0].desc = MKLDNNMemoryDesc(getParentEdgeAt(0)->getDims(), dataType, memoryFormat);
+        config.outConfs[0].desc = MKLDNNMemoryDesc(getChildEdgeAt(0)->getDims(), dataType, memoryFormat);
+        supportedPrimitiveDescriptors.push_back({config, impl_desc_type::ref, memoryFormat});
```
Thank you! Done

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -82,7 +85,111 @@ void MKLDNNPadNode::initSupportedPrimitiveDescriptors() {
+    }
+}
+
+void MKLDNNPadNode::selectOptimalPrimitiveDescriptor() {
```
Only blocked layouts for Constant mode support case when paddings for channel are % 8[16] = 0. Another modes support cases when  paddings for channel are 0. So for these modes, if parent has blocked layout, `n[d]hwc` must be selected

dmitry-gorokhov replied:
Based on your description I would say that all you need to do is to create supportedPrimitiveDescriptor for n[d]hwc format before nc[d]hw. In that case if parent returns blocked layout and Pad doesn't support it then default selectOptimalPrimitiveDescriptor routine will choose first from the supported array - n[d]hwc.

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -95,158 +202,312 @@ void MKLDNNPadNode::createPrimitive() {
+        padsEnd[1] /= params.srcDims[params.srcDims.size() - 1];
+        padsBegin.push_back(0);
+        padsEnd.push_back(0);
+    } else if (layout == NHWC || layout == NDHWC) {
```
Thank you! Done

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -95,158 +202,312 @@ void MKLDNNPadNode::createPrimitive() {
-            getParentEdgeAt(0)->getBlob()->getTensorDesc().getBlockingDesc().getOffsetPadding();
-    float* dstData = getChildEdgeAt(0)->getBlob()->buffer().as<float*>() +
-            getChildEdgeAt(0)->getBlob()->getTensorDesc().getBlockingDesc().getOffsetPadding();
+void MKLDNNPadNode::prepareParams() {
```
Ok, I put it in one place

a-sidorova commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -95,158 +202,312 @@ void MKLDNNPadNode::createPrimitive() {
+              OV_CASE(InferenceEngine::Precision::I16, int16_t),
+              OV_CASE(InferenceEngine::Precision::U16, uint16_t),
+              OV_CASE(InferenceEngine::Precision::I8, int8_t),
+              OV_CASE(InferenceEngine::Precision::U8, uint8_t));
```
Thank you. Removed

dmitry-gorokhov commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -80,9 +88,94 @@ void MKLDNNPadNode::initSupportedPrimitiveDescriptors() {
+    };
+    pushSupportedPrimitiveDescriptor(memoryFormat);
+
+    if (numOfDims != 4 && numOfDims != 5)
```
Seems like we don't need this condition at all

dmitry-gorokhov commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -80,9 +88,94 @@ void MKLDNNPadNode::initSupportedPrimitiveDescriptors() {
+        config.outConfs[0].desc = MKLDNNMemoryDesc(getChildEdgeAt(0)->getDims(), dataType, memoryFormat);
+        supportedPrimitiveDescriptors.push_back({config, impl_desc_type::ref, memoryFormat});
+    };
+    pushSupportedPrimitiveDescriptor(memoryFormat);
```
Do we need explicit `memoryFormat` variable? Lets write stm like: pushSupportedPrimitiveDescriptor(MKLDNNMemory::GetPlainFormat(getParentEdgeAt(0)->getDims()));

dmitry-gorokhov commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -80,9 +88,94 @@ void MKLDNNPadNode::initSupportedPrimitiveDescriptors() {
+    if (parent->getSelectedPrimitiveDescriptor() != nullptr) {
+        int outputIndex = getParentEdgeAt(0)->getOutputNum();
+        if (outputIndex < 0)
+            THROW_IE_EXCEPTION << "Cannot find index of output node";
```
BTW, the same comment as for another review: https://github.com/openvinotoolkit/openvino/pull/3449#discussion_r539979404

a-sidorova commented:
@dmitry-gorokhov, thank you for review! I fixed all.

I changed logic in `initSupportedPrimitiveDescriptors` - first descriptor with n[d]hwc is created, then descriptor with planar layout is created and in the end if possible, then descriptor with blocked layout is created. So I removed separate `supportSupportedPrimitiveDescriptors`  for Pad.

And I added Pad node name in error messages.

dmitry-gorokhov commented:
@a-sidorova Related PR: https://github.com/openvinotoolkit/openvino/pull/3449 was merged. Please rebase your branch on latest master since we need to recollect numbers with both patches enabled.

a-sidorova commented:
@dmitry-gorokhov, Done

dmitry-gorokhov reviewed (approved):
Perf results looks good. I need green CI to merge the PR.

maxnick commented on inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp:
```diff
@@ -79,174 +87,347 @@ void MKLDNNPadNode::initSupportedPrimitiveDescriptors() {
+
+inline uint8_t* MKLDNNPadNode::getDataPtr(const MKLDNNMemory& memoryPtr) const {
+    return reinterpret_cast<uint8_t*>(memoryPtr.GetData()) + memoryPtr.GetDescriptor().data.layout_desc.blocking.offset_padding *
+                                      MKLDNNExtensionUtils::sizeOfDataType(mkldnn::memory::data_type(memoryPtr.GetDescriptor().data.data_type));
```
The offset_padding has an ptrdiff_t type, so it seems to me that there is no need to multiply this value by the element size since this is an already calculated offset to the beginning address. One can check MKLDNNMemory::SetData method to confirm my statement.

a-sidorova replied:
You're right. Thank you!

maxnick replied:
Sorry, I was wrong. Since we calculate offset with int8 ptrs, we have to multiply the offset by the data size, because ptr arithmetic implies address calculation in data blocks with the size of data type, and actually the offset_padding is in the data blocks but not in bytes.

a-sidorova replied:
Ok, I returned back

a-sidorova commented:
@dmitry-gorokhov, done
